ethereum/core/src/main/java/org/hyperledger/besu/ethereum/storage/keyvalue/WorldStateKeyValueStorage.java

```diff
@@ -86,9 +89,32 @@ public Updater updater() {
     return new Updater(keyValueStorage.startTransaction(), nodeAddedListeners);
   }
 
+  static final AtomicReference<Optional<byte[]>> doomedKeyRef =
+      new AtomicReference<>(Optional.empty());
+
   @Override
   public long prune(final Predicate<byte[]> inUseCheck) {
-    return keyValueStorage.removeAllKeysUnless(inUseCheck);
+    final AtomicInteger prunedKeys = new AtomicInteger(0);
+    try (final Stream<byte[]> keys = keyValueStorage.streamKeys()) {
+      keys.forEach(
+          key -> {
+            doomedKeyRef.set(Optional.of(key));
+            if (!inUseCheck.test(key)) {
+              // even though we just set this above, we might unset it in commit below to prevent
+              // thread interleaving from causing an erroneous deletion
+              doomedKeyRef
+                  .get()
+                  .ifPresent(
+                      doomedKey -> {
+                        keyValueStorage.delete(doomedKey);
+                        prunedKeys.incrementAndGet();
+                      });
+            }
+            doomedKeyRef.set(Optional.empty());
```

Comment:
I can't work out how this helps with the threading issue.  The bug would be a lot less likely to occur because there's now a second check done but I don't see what's guaranteeing the threading order to guarantee that a transaction isn't adding the key between when we do `doomedKeyReft.get().ifPresent` and the actual deletion of the key.

Is there some detail of RocksDB threading and locking I'm missing?

Reply:
Nope, you're right. I just brain farted.